modal/io_streams.py

```diff
@@ -116,7 +118,7 @@ def __init__(
     def file_descriptor(self):
         return self._file_descriptor
 
-    async def read(self) -> str:
+    async def read(self) -> Union[str, bytes]:
```

Comment:
Having a union type for the return type here would make type-checkers complain whenever someones uses this API without checking the return type.

Type safety could be solved by making `_StreamReader` into a `typing.Generic[T]` with T determined by which `text` mode it uses, but considering that our transfer protocol is already a "string" and all we do with `text=False` is re-encode it as hard-coded utf8, I'd lean towards skipping the `text=False` option entirely and let users do that encoding themselves if they really want bytes?

Somewhat unrealted - but it feels somewhat upside down that we use `str` as the basic transfer protocol for these "logs" instead of `bytes` - what happens if the output string of a command isn't a valid utf-8 string for example? Would that cause an error? And if so, `text=False` is a bit misleading for users too...

Reply:
I agree! I'll look into changing the transfer protocol before merging this change. I think we'd like to keep the interface close to `subprocess.Popen`, which was the original incentive for adding `text`.